cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,346 @@
+
+  if (p1_idx >= linestring1_num_points) { return; }
+
+  auto const linestring_idx =
+    thrust::distance(linestring1_offsets_begin,
+                     thrust::upper_bound(
+                       thrust::seq, linestring1_offsets_begin, linestring1_offsets_end, p1_idx)) -
+    1;
+
+  auto ls1_end = endpoint_index_of_linestring(
+    linestring_idx, linestring1_offsets_begin, num_linestrings, linestring1_num_points);
+
+  if (p1_idx == ls1_end) {
+    // Current point is the end point of the line string.
+    return;
+  }
+
+  auto ls2_start = *(linestring2_offsets_begin + linestring_idx);
+  auto ls2_end   = endpoint_index_of_linestring(
+    linestring_idx, linestring2_offsets_begin, num_linestrings, linestring2_num_points);
+
+  auto const& A = thrust::raw_reference_cast(linestring1_points_begin[p1_idx]);
+  auto const& B = thrust::raw_reference_cast(linestring1_points_begin[p1_idx + 1]);
+
+  auto min_squared_distance = std::numeric_limits<T>::max();
+  for (auto p2_idx = ls2_start; p2_idx < ls2_end; p2_idx++) {
+    auto const& C        = thrust::raw_reference_cast(linestring2_points_begin[p2_idx]);
+    auto const& D        = thrust::raw_reference_cast(linestring2_points_begin[p2_idx + 1]);
+    min_squared_distance = std::min(min_squared_distance, squared_segment_distance(A, B, C, D));
+  }
+
+  atomicMin(&thrust::raw_reference_cast(*(distances + linestring_idx)),
+            static_cast<T>(std::sqrt(min_squared_distance)));
+}
+
+}  // namespace detail
+
+template <class Cart2dItA,
+          class Cart2dItB,
+          class OffsetIterator,
+          class OutputIt,
+          class Cart2dA,
+          class Cart2dB>
+void pairwise_linestring_distance(OffsetIterator linestring1_offsets_first,
+                                  OffsetIterator linestring1_offsets_last,
+                                  Cart2dItA linestring1_points_first,
+                                  Cart2dItA linestring1_points_last,
+                                  OffsetIterator linestring2_offsets_first,
+                                  Cart2dItB linestring2_points_first,
+                                  Cart2dItB linestring2_points_last,
+                                  OutputIt distances_first,
+                                  rmm::cuda_stream_view stream)
+{
+  using T = typename Cart2dA::value_type;
+
+  static_assert(detail::is_floating_point<T,
+                                          typename Cart2dB::value_type,
+                                          typename std::iterator_traits<OutputIt>::value_type>(),
+                "Inputs and output must be floating point types.");
+
+  static_assert(detail::is_same<T,
+                                typename Cart2dB::value_type,
+                                typename std::iterator_traits<OutputIt>::value_type>(),
+                "Inputs and output must be the same types.");
+
+  static_assert(detail::is_same<cartesian_2d<T>, Cart2dA, Cart2dB>(),
```

Comment:
If this is the case, then you don't need both Cart2dA and Cart2dB -- just use the same type for both inputs.

Reply:
Unfortunately `Cart2dA` and `Cart2dB` are just two alias for the `value_type` of iterator `Cart2dItA` and `Cart2dItB`, note that they are parameters derived from existing template parameters. The test here is basically asserting that the value type for both iterators should be `cartesian_2d<T>`.

Reply:
I get that. But you only need one type parameter if they are both required to be the same type, right? See how I did it for haversine.  Single parameter for `Location` type:
https://github.com/rapidsai/cuspatial/blob/1f070d2589b7a4fb6d5299762630d6d4f9b843ae/cpp/include/cuspatial/experimental/haversine.cuh#L71-L81

Then check that both `Location` and the `value_type` of the B iterator are `lonlat_2d`:
https://github.com/rapidsai/cuspatial/blob/1f070d2589b7a4fb6d5299762630d6d4f9b843ae/cpp/include/cuspatial/experimental/detail/haversine.cuh#L70-L73